### PR TITLE
WDP-28

### PR DESCRIPTION
**Pull request by KamiNoSierhej**
https://kindergarten.stxnext.pl/browse/WDP-28

### CODE REVIEW

**Comment by KamiNoSierhej** [comment]
https://kindergarten.stxnext.pl/browse/WDP-28

File: src/presence_analyzer/static/presence_start_end.html [hunk 1]
```diff
@@ -96,15 +100,12 @@
...
-                            ["Wed", new Date(1, 1, 1, 7, 00, 06), new Date(1, 1, 1, 13, 24, 12)],
-                            ["Thu", new Date(1, 1, 1, 7, 26, 32), new Date(1, 1, 1, 12, 40, 45)],
-                            ["Fri", new Date(1, 1, 1, 6, 53, 22), new Date(1, 1, 1, 12, 52, 09)]
-                        ] 
+                        $.getJSON("/api/v1/presence_start_end/"+selected_user, function(result) {
+                        $.each(result, function(index, value) {
+                                value[1] = parseInterval(value[1]);
+                                value[2] = parseInterval(value[2]);
+                            });
+                            console.log(result);
```
Comment (dpolcyn) [inline src/presence_analyzer/static/presence_start_end.html]: Why you added it here?

File: src/presence_analyzer/views.py [hunk 2]
```diff
@@ -2,12 +2,18 @@
...
 """
-
+import datetime
+import time
+import json
 import calendar
 from flask import redirect, abort
 
 from main import app
+<<<<<<< 26fcd6a738da14e7892f8c2478da5802ea502915
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: ???

File: src/presence_analyzer/utils.py [hunk 3]
```diff
@@ -103,3 +102,14 @@ def mean(items):
     Calculates arithmetic mean. Returns zero for empty lists.
     """
     return float(sum(items)) / len(items) if len(items) > 0 else 0
+
+def day_start_end(items):
+    """
+    Groups times of start and end work at weekday.
+    """
+    result = [[], [], [], [], [], [], []]
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]: Use list comprehension

File: src/presence_analyzer/views.py [hunk 4]
```diff
@@ -72,3 +78,32 @@ def presence_weekday_view(user_id):
...
+
+    data = get_data()
+    if user_id not in data:
+        log.debug('User %s not found!', user_id)
+        abort(404)
+
+    startend = day_start_end(data[user_id])
+    result = []
+    days = calendar.day_abbr
+    day_num = 0
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: I don't see reason of declaration this value.

File: src/presence_analyzer/tests.py [hunk 5]
```diff
@@ -105,6 +105,49 @@ def test_mean_time_weekday_view(self):
             ])
         )
 
+    def test_mean_time_weekday_view(self):
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Why did you add this test?

File: src/presence_analyzer/static/presence_start_end.html (hunk 1 above)
Comment (KamiNoSierhej) [inline src/presence_analyzer/static/presence_start_end.html]: For debug and then just forgot about that

File: src/presence_analyzer/views.py (hunk 2 above)
Comment (KamiNoSierhej) [inline src/presence_analyzer/views.py]: I was doing git --rebase and just don't noticed i left that

File: src/presence_analyzer/views.py (hunk 4 above)
Comment (KamiNoSierhej) [inline src/presence_analyzer/views.py]: for item in startend[day_num]: <<< I using this here so i declarated to avoid referenced before assignment.

File: src/presence_analyzer/tests.py (hunk 5 above)
Comment (KamiNoSierhej) [inline src/presence_analyzer/tests.py]: Because of --rebase some of things get multiplied and i just don't noticed these one.

File: src/presence_analyzer/static/presence_start_end.html [hunk 6]
```diff
@@ -96,15 +101,11 @@
...
-                        var result = [
-                            ["Mon", new Date(1, 1, 1, 7, 37, 38), new Date(1, 1, 1, 15, 07, 14)],
-                            ["Tue", new Date(1, 1, 1, 8, 16, 23), new Date(1, 1, 1, 14, 01, 47)],
-                            ["Wed", new Date(1, 1, 1, 7, 00, 06), new Date(1, 1, 1, 13, 24, 12)],
-                            ["Thu", new Date(1, 1, 1, 7, 26, 32), new Date(1, 1, 1, 12, 40, 45)],
-                            ["Fri", new Date(1, 1, 1, 6, 53, 22), new Date(1, 1, 1, 12, 52, 09)]
-                        ] 
+                        $.getJSON("/api/v1/presence_start_end/"+selected_user, function(result) {
+                        $.each(result, function(index, value) {
+                                value[1] = parseInterval(value[1]);
```
Comment (dpolcyn) [inline src/presence_analyzer/static/presence_start_end.html]: fix style

File: src/presence_analyzer/views.py [hunk 7]
```diff
@@ -72,3 +75,19 @@ def presence_weekday_view(user_id):
...
     return result
+
+
+@app.route('/api/v1/presence_start_end/<int:user_id>', methods=['GET'])
+@jsonify
+def presence_start_end(user_id):
+    """
+    The medium time to come to the office and medium time of leave.
+    """
+
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: One empty line.

File: src/presence_analyzer/views.py [hunk 8]
```diff
@@ -2,12 +2,15 @@
...
+import datetime
+import time
+import json
 import calendar
-from flask import redirect, abort
 
+from flask import redirect, abort
 from main import app
-from utils import get_data, group_by_weekday, jsonify, mean
+from utils import day_start_end, get_data, group_by_weekday, seconds_since_midnight, jsonify, mean
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: Sort imports

**Comment by dpolcyn** [comment]
Why you made changes in src/presence_analyzer/static/js/jquery.min.js?

**Comment by KamiNoSierhej** [comment]
Just reparation of outdated things nothing special.

File: src/presence_analyzer/tests.py [hunk 9]
```diff
@@ -105,6 +104,26 @@ def test_mean_time_weekday_view(self):
...
+    def test_presence_start_end(self):
+        """
+        Test the medium time to come to the office and medium time of leave.
+        """
+        resp = self.client.get('/api/v1/presence_start_end/10')
+        self.assertEqual(resp.status_code, 200)
+        self.assertEqual(resp.content_type, 'application/json')
+        data = json.loads(resp.data)
+        self.assertEqual(
+            data,
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: data is a tuple?

File: src/presence_analyzer/static/presence_start_end.html [hunk 10]
```diff
@@ -96,33 +101,28 @@
...
+                            $.each(result, function(index, value) {
+                                value[1] = parseInterval(value[1]);
+                                value[2] = parseInterval(value[2]);
+                            });
+                            var data = new google.visualization.DataTable();
+                            data.addColumn('string', 'Weekday');
+                            data.addColumn({ type: 'datetime', id: 'Start' });
+                            data.addColumn({ type: 'datetime', id: 'End' });
+                            data.addRows(result);
+                            var options = {
```
Comment (dpolcyn) [inline src/presence_analyzer/static/presence_start_end.html]: How to declare multiple variables in js:
```
var options = {
    a = 23131,
    b = 234342,
},
    dlasda = 54353;
```

File: src/presence_analyzer/static/presence_start_end.html [hunk 11]
```diff
@@ -79,6 +79,11 @@
         google.load("visualization", "1", {packages:["corechart", "timeline"], 'language': 'pl'});
     </script>
     <script type="text/javascript">
+        function parseInterval(value) {
```
Comment (dpolcyn) [inline src/presence_analyzer/static/presence_start_end.html]: Duplication, you already have this function in another file.

File: src/presence_analyzer/static/js/parse.js [hunk 12]
```diff
@@ -0,0 +1,5 @@
+function parseInterval(value) {
```
Comment (dpolcyn) [inline src/presence_analyzer/static/js/parse.js]: Parse? It can be any parse function?

File: src/presence_analyzer/static/js/parse.js [hunk 13]
```diff
@@ -0,0 +1,5 @@
...
+    var result = new Date(1,1,1);
+    result.setMilliseconds(value*1000);
+    return result;
+}
```
Comment (dpolcyn) [inline src/presence_analyzer/static/js/parse.js]: Empty line on the end of file.

File: src/presence_analyzer/static/mean_time_weekday.html [hunk 14]
```diff
@@ -113,11 +112,11 @@
                             data.addRows(result);
                             var options = {
                                 hAxis: {title: 'Weekday'}
-                            };
-                            var formatter = new google.visualization.DateFormat({pattern: 'HH:mm:ss'});
+                            },
+                            formatter = new google.visualization.DateFormat({pattern: 'HH:mm:ss'});
```
Comment (dpolcyn) [inline src/presence_analyzer/static/mean_time_weekday.html]: Indentation. I gave you example code before.

File: src/presence_analyzer/static/presence_start_end.html [hunk 15]
```diff
@@ -74,11 +74,20 @@
...
     <script type="text/javascript" src="https://www.google.com/jsapi"></script>
     <script type="text/javascript">
         google.load("visualization", "1", {packages:["corechart", "timeline"], 'language': 'pl'});
     </script>
-    <script type="text/javascript">
+    <script>
+        parseInterval();
+    </script>
+    <script type="text/javascript">/*
+        function parseInterval(value) {
```
Comment (dpolcyn) [inline src/presence_analyzer/static/presence_start_end.html]: This code is necessary?

File: src/presence_analyzer/views.py [hunk 16]
```diff
@@ -72,3 +75,18 @@ def presence_weekday_view(user_id):
...
+    """
+    The medium time to come to the office and medium time of leave.
+    """
+    data = get_data()
+    if user_id not in data:
+        log.debug('User %s not found!', user_id)
+        abort(404)
+
+    result = day_start_end(data[user_id])
+    return result
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: return day_start_end(data[user_id])?
